[Bugfix] Schedule failure due to wrong get_image_size_with_most_features

tomtomjhj opened the pull request:
## Purpose

The scheduler may fail to schedule a batch if `get_image_size_with_most_features` is wrong and `max_num_batched_tokens` is not big enough:
* Some models' `get_image_size_with_most_features` returns an image size that is encoded into a sequence shorter than the actual maximum. That is, `max_tokens_per_mm_item < num_mm_input_tokens` is possible.
* `encoder_compute_budget` is the max of the sequence length of an image with size `get_image_size_with_most_features` and `max_num_batched_tokens` (= `max_num_encoder_input_tokens`). <https://github.com/vllm-project/vllm/blob/0808eb813b6b219324092ab8cc25d3223e5ccb77/vllm/v1/core/encoder_cache_manager.py#L336-L338>
* Therefore, if `max_num_batched_tokens < num_mm_input_tokens`, it is possible that `max_num_encoder_input_tokens < num_mm_input_tokens`, which results in scheduling failure. <https://github.com/vllm-project/vllm/blob/0808eb813b6b219324092ab8cc25d3223e5ccb77/vllm/v1/core/encoder_cache_manager.py#L139-L141>

This PR fixes `get_image_size_with_most_features` of qwen2_vl and gemma3.
* qwen2_vl
    * Problem: An image with with most features is not a square if the number of features is not a squre.
    * Solution: Factorize the max number of features and contruct the image shape that is closest to square to meet the aspect ratio constraint.
* gemma3
    * Problem: The image is not big enough to trigger cropping.
    * Solution: Use the native image size.


## Test Plan
Apply the following patch
```diff
diff --git a/examples/offline_inference/vision_language.py b/examples/offline_inference/vision_language.py
index 8f72bf6f0..755aba99d 100755
--- a/examples/offline_inference/vision_language.py
+++ b/examples/offline_inference/vision_language.py
@@ -321,6 +321,7 @@ def run_gemma3(questions: list[str], modality: str) -> ModelRequestData:
         max_num_seqs=2,
         mm_processor_kwargs={"do_pan_and_scan": True},
         limit_mm_per_prompt={modality: 1},
+        max_num_batched_tokens=512,
     )
 
     prompts = [
@@ -1527,6 +1528,7 @@ def run_qwen2_5_vl(questions: list[str], modality: str) -> ModelRequestData:
             "fps": 1,
         },
         limit_mm_per_prompt={modality: 1},
+        max_num_batched_tokens=512,
     )
 
     if modality == "image":
```

Then run
```
python examples/offline_inference/vision_language.py --model-type qwen2_5_vl
```
and
```
python examples/offline_inference/vision_language.py --model-type gemma3
```


## Test Result
Without this PR's fix, vLLM will hang due to repeated scheduler failure. This PR fixes this problem.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

tomtomjhj committed:
fix: qwen2_vl image with most features

Problem:
An image with with most features is not a square if the number of
features is not a squre.

Solution:
Factorize the max number of features and contruct the image shape that
is closest to square to meet the aspect ratio constraint.

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

tomtomjhj committed:
fix: gemma3 image with most features

Problem:
The image is not big enough to trigger cropping.

Solution:
Use the native image size.

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

DarkLight1337 reviewed (commented):
Thanks for fixing, can you add some regression tests using the image size that causes this failure?

tomtomjhj committed:
tests

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

tomtomjhj commented:
Added unit tests for `get_image_size_with_most_features`. These tests fail without my fix.

tomtomjhj committed:
fix: aspect ratio constraint

If the seq len can't be factored into a pair that satifies the aspect
ratio contraint, decrement the seq len and retry.

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

ywang96 commented:
Looks like there are some test failures

```
unsupported operand type(s) for //: 'NoneType' and 'int'
```

tomtomjhj commented on vllm/model_executor/models/qwen2_vl.py:
```diff
@@ -959,13 +960,29 @@ def get_num_video_tokens(
+        for seq_len in range(max_seq_len, 0, -1):
+            height_factor, width_factor = closest_factor_pair(seq_len)
+            if width_factor / height_factor <= 200:
+                break
```
I think `_get_vision_info` is doing its job correctly. Qwen2 processes huge square image into 35x35 patches when given pixel limit 1280x28x28, so `_get_vision_info` should follow that behavior.

ywang96 replied:
@tomtomjhj Thanks for your explaination and I think I understand what the problem is - here we're really trying to find an image size that would almost saturate the 1280x28x28 max pixels when an image of `999999x999999` cannot because it's square size. For that reason I think this fix makes sense.

I think it would be great if we can document the context here why we're not leveraging `_get_vision_info`!

tomtomjhj committed:
fix: max_pixels can be None in qwen3_vl

Fallback to "longest_edge".

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

tomtomjhj committed:
comments

Signed-off-by: Jaehwang Jung <tomtomjhj@gmail.com>

ywang96 committed:
Merge branch 'main' into image-with-most-features

tomtomjhj committed:
Merge branch 'main' into image-with-most-features
